File: solutions/haiku-1/haiku/typescript/brackets.ts
import { createInterface } from "readline";
import { stdin } from "process";

function isBalanced(line: string): boolean {
  const stack: string[] = [];

  for (const char of line) {
    if (char === "(" || char === "[" || char === "{") {
      stack.push(char);
    } else if (char === ")" || char === "]" || char === "}") {
      if (stack.length === 0) {
        return false;
      }

      const opening = stack.pop()!;

      if (
        (char === ")" && opening !== "(") ||
        (char === "]" && opening !== "[") ||
        (char === "}" && opening !== "{")
      ) {
        return false;
      }
    }
  }

  return stack.length === 0;
}

const rl = createInterface({
  input: stdin,
  terminal: false,
});

rl.on("line", (line: string) => {
  console.log(isBalanced(line) ? "yes" : "no");
});

File: solutions/haiku-1/haiku/java/brackets.java
import java.util.Scanner;
import java.util.Stack;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        while (scanner.hasNextLine()) {
            String line = scanner.nextLine();

            if (isBalanced(line)) {
                System.out.println("yes");
            } else {
                System.out.println("no");
            }
        }

        scanner.close();
    }

    public static boolean isBalanced(String line) {
        Stack<Character> stack = new Stack<>();

        for (char c : line.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else if (c == ')' || c == ']' || c == '}') {
                if (stack.isEmpty()) {
                    return false;
                }

                char opening = stack.pop();

                if (!isMatchingPair(opening, c)) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }

    public static boolean isMatchingPair(char opening, char closing) {
        return (opening == '(' && closing == ')') ||
               (opening == '[' && closing == ']') ||
               (opening == '{' && closing == '}');
    }
}
